### PR TITLE
Share collections

### PR DESCRIPTION
**Pull request by sharwell**
* Create a shared project **Microsoft.CodeAnalysis.Collections** for shared collections, and make it available in both the compiler and workspace layers
* Move `SegmentedArray<T>` to **Microsoft.CodeAnalysis.Collections**

### CODE REVIEW

File: src/Dependencies/Collections/Microsoft.CodeAnalysis.Collections.projitems [hunk 1]
```diff
@@ -0,0 +1,22 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (jaredpar) [inline src/Dependencies/Collections/Microsoft.CodeAnalysis.Collections.projitems:1]: Consider adding a README.md file here that explains what the purpose of this component is and why developers may want to leverage it.

File: Roslyn.sln [hunk 2]
```diff
@@ -475,6 +475,10 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.CodeAnalysis.Test
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.CodeAnalysis.XunitHook", "src\EditorFeatures\XunitHook\Microsoft.CodeAnalysis.XunitHook.csproj", "{967A8F5E-7D18-436C-97ED-1DB303FE5DE0}"
 EndProject
+Project("{D954291E-2A0B-460D-934E-DC6B0785DB48}") = "Microsoft.CodeAnalysis.Collections", "src\Dependencies\Collections\Microsoft.CodeAnalysis.Collections.shproj", "{E919DD77-34F8-4F57-8058-4D3FF4C2B241}"
```
Comment (jaredpar) [inline Roslyn.sln:478]: Need a similar update for Compilers.sln

File: src/Dependencies/Collections/Microsoft.CodeAnalysis.Collections.shproj [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+<!-- Licensed to the .NET Foundation under one or more agreements. The .NET Foundation licenses this file to you under the MIT license. See the LICENSE file in the project root for more information. -->
+<Project>
+  <PropertyGroup Label="Globals">
+    <ProjectGuid>e919dd77-34f8-4f57-8058-4d3ff4c2b241</ProjectGuid>
+    <MinimumVisualStudioVersion>14.0</MinimumVisualStudioVersion>
+  </PropertyGroup>
+  <Import Project="$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props" Condition="Exists('$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props')" />
+  <Import Project="$(MSBuildExtensionsPath32)\Microsoft\VisualStudio\v$(VisualStudioVersion)\CodeSharing\Microsoft.CodeSharing.Common.Default.props" Condition="Exists('$(MSBuildExtensionsPath32)\Microsoft\VisualStudio\v$(VisualStudioVersion)\CodeSharing\Microsoft.CodeSharing.Common.Default.props')" />
+  <Import Project="$(MSBuildExtensionsPath32)\Microsoft\VisualStudio\v$(VisualStudioVersion)\CodeSharing\Microsoft.CodeSharing.Common.props" Condition="Exists('$(MSBuildExtensionsPath32)\Microsoft\VisualStudio\v$(VisualStudioVersion)\CodeSharing\Microsoft.CodeSharing.Common.props')" />
```
Comment (jaredpar) [inline src/Dependencies/Collections/Microsoft.CodeAnalysis.Collections.shproj:10]:
```suggestion
    <Import Project="$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props" Condition="Exists('$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props')" />
    <Import Project="$(MSBuildExtensionsPath32)\Microsoft\VisualStudio\v$(VisualStudioVersion)\CodeSharing\Microsoft.CodeSharing.Common.Default.props" Condition="Exists('$(MSBuildExtensionsPath32)\Microsoft\VisualStudio\v$(VisualStudioVersion)\CodeSharing\Microsoft.CodeSharing.Common.Default.props')" />
    <Import Project="$(MSBuildExtensionsPath32)\Microsoft\VisualStudio\v$(VisualStudioVersion)\CodeSharing\Microsoft.CodeSharing.Common.props" Condition="Exists('$(MSBuildExtensionsPath32)\Microsoft\VisualStudio\v$(VisualStudioVersion)\CodeSharing\Microsoft.CodeSharing.Common.props')" />
```

File: src/Dependencies/Collections/Microsoft.CodeAnalysis.Collections.shproj [hunk 4]
```diff
@@ -0,0 +1,14 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<!-- Licensed to the .NET Foundation under one or more agreements. The .NET Foundation licenses this file to you under the MIT license. See the LICENSE file in the project root for more information. -->
+<Project>
+  <PropertyGroup Label="Globals">
+    <ProjectGuid>e919dd77-34f8-4f57-8058-4d3ff4c2b241</ProjectGuid>
+    <MinimumVisualStudioVersion>14.0</MinimumVisualStudioVersion>
```
Comment (jaredpar) [inline src/Dependencies/Collections/Microsoft.CodeAnalysis.Collections.shproj:6]: Do we need this property anymore?

**Review by jaredpar — CHANGES_REQUESTED** [review]

**Review by 333fred — COMMENTED** [review]
I like the general idea. Do you think it would be a good idea to centralize all of our `ImmutableArray` and other collection extension methods in this project as well?

File: src/Dependencies/Collections/Microsoft.CodeAnalysis.Collections.shproj (hunk 4 above)
Comment (sharwell) [reply src/Dependencies/Collections/Microsoft.CodeAnalysis.Collections.shproj:6]: I'm not sure. I just copied it from other shared projects we have.

File: src/Dependencies/Collections/Microsoft.CodeAnalysis.Collections.shproj (hunk 3 above)
Comment (sharwell) [reply src/Dependencies/Collections/Microsoft.CodeAnalysis.Collections.shproj:10]: ➡️ These appear to already be indented correctly.

File: src/Dependencies/Collections/Microsoft.CodeAnalysis.Collections.projitems (hunk 1 above)
Comment (sharwell) [reply src/Dependencies/Collections/Microsoft.CodeAnalysis.Collections.projitems:1]: ➡️ Added a simple README file

File: Roslyn.sln (hunk 2 above)
Comment (sharwell) [reply Roslyn.sln:478]: ➡️ Updated both **Compilers.sln** and **Compilers.slnf**.

File: src/Dependencies/Collections/SegmentedArray`1.cs [hunk 5]
```diff
@@ -71,7 +70,7 @@ public SegmentedArray(int length)
                 // Avoid using (length & s_offsetMask) because it doesn't handle a last page size of s_segmentSize.
                 var lastPageSize = length - ((_items.Length - 1) << s_segmentShift);
 
-                _items[^1] = new T[lastPageSize];
+                _items[_items.Length - 1] = new T[lastPageSize];
```
Comment (333fred) [inline src/Dependencies/Collections/SegmentedArray`1.cs:73]: What's the purpose of this change? Does it just not have the range types available?

Comment (sharwell) [reply src/Dependencies/Collections/SegmentedArray`1.cs:73]: Correct. It's easy to assume the files are provided in dotnet/roslyn, but we can't make the same assumption if the shared project is consumed by another repository.

**Comment by sharwell** [comment]
> Do you think it would be a good idea to centralize all of our `ImmutableArray` and other collection extension methods in this project as well?

The code I currently have planned for this closely matches APIs that exist in the BCL. Since this shared project is intended to be consumed by other repositories, I don't want to make too many assumptions about the usefulness of extensions to the standard APIs.

File: src/Dependencies/Collections/SegmentedArray`1.cs (hunk 5 above)
Comment (333fred) [reply src/Dependencies/Collections/SegmentedArray`1.cs:73]: > but we can't make the same assumption if the shared project is consumed by another repository.

Is that a goal of this change? If so, why?

Comment (sharwell) [reply src/Dependencies/Collections/SegmentedArray`1.cs:73]: I do make the assumption that **NullableAnnotations.cs** is provided. This seems to be a common expectation.

Comment (333fred) [reply src/Dependencies/Collections/SegmentedArray`1.cs:73]: What about the more general question though? Who is going to be consuming _source code_ from the dotnet/roslyn repository, and why is that a goal for this change? It seems to me that that would be an anti-goal: it's like an IVT, but without even the advertisement of an `InternalsVisibleTo` attribute.

Comment (sharwell) [reply src/Dependencies/Collections/SegmentedArray`1.cs:73]: Current candidates are dotnet/project-system, CPS, and MSBuild.

File: src/Dependencies/Collections/SegmentedArray`1.cs [hunk 6]
```diff
@@ -7,7 +7,6 @@
 using System.Collections.Generic;
 using System.Reflection.Emit;
 using System.Runtime.CompilerServices;
-using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.Shared.Collections
```
Comment (tmat) [inline src/Dependencies/Collections/SegmentedArray`1.cs:11]: >Shared [](start = 33, length = 6)

I'd omit the "Shared" subnamespace. It doesn't add much value, imo.

File: src/Dependencies/Collections/SegmentedArray`1.cs (hunk 5 above)
Comment (tmat) [reply src/Dependencies/Collections/SegmentedArray`1.cs:73]: We already use source packages to share Debugging helpers and PooledObjects. Source packages do not have the same issues as IVTs. We can make any changes, even breaking without breaking anyone. This is because the project that uses the source package ends up having their own copy of the source that won't get updated unless the project explicitly does so and builds in their CI. 

---
In reply to: [543702405](https://github.com/dotnet/roslyn/pull/49980#discussion_r543702405) [](ancestors = 543702405)

**Comment by tmat** [comment]
LGTM, other than the "Shared" subnamespace. I'd remove it.

File: src/Dependencies/Collections/README.md [hunk 7]
```diff
@@ -0,0 +1,5 @@
+﻿# Microsoft.CodeAnalysis.Collections
+
+This package contains shared code implementing specialized collection types.
+
+* `SegmentedArray<T>`: This type behaves similarly to `T[]`, but segments the underlying storage to avoid allocating arrays in the Large Object Heap.
```
Comment (jaredpar) [inline src/Dependencies/Collections/README.md:5]: Is there any good advice on when to use one over the other that you could include / link to here ?

**Review by jaredpar — APPROVED** [review]

File: src/Dependencies/Collections/README.md (hunk 7 above)
Comment (sharwell) [reply src/Dependencies/Collections/README.md:5]: The long form might look something like this:

> If your code relies on the indexer performance characteristics of `T[]`, does not rely on `AsSpan()`/`AsMemory()`, and performance measurements show problematic results from Large Object Heap allocations, it may benefit from using `SegmentedArray<T>` instead of `T[]`.

It seems likely that once a user knows that Large Object Heap is a problem that needs to be avoided, the statement already in the README would be sufficient to help them in determining the correct type to use.

All the types I currently have planned for this package are highly application-specific and motivated by profiling results.

**Review by tmat — APPROVED** [review]

File: src/Dependencies/Collections/README.md (hunk 7 above)
Comment (jaredpar) [reply src/Dependencies/Collections/README.md:5]: > All the types I currently have planned for this package are highly application-specific and motivated by profiling results.

Seems like a good bit to include. For example linking to a PR or two where you made conversions to `SegmentedArray<T>` due to profiling pushing you in that direction. Particularly if the PR description has data in it then it would serve as a nice "treasure map" for other developers to follow. 

> It seems likely that once a user knows that Large Object Heap is a problem that needs to be avoided, the statement already in the README would be sufficient to help them in determining the correct type to use.

Think you're applying too much of your internal understanding of performance and this type to the answer here. Think a good chunk of devs would find this type who don't quite have the understanding level you're expecting here.

File: src/Dependencies/Collections/SegmentedArray`1.cs (hunk 6 above)
Comment (sharwell) [reply src/Dependencies/Collections/SegmentedArray`1.cs]: ➡️ Removed

File: src/Dependencies/Collections/README.md (hunk 7 above)
Comment (sharwell) [reply src/Dependencies/Collections/README.md]: 👍 Added some details

File: src/Dependencies/Collections/README.md [hunk 8]
```diff
@@ -0,0 +1,11 @@
+﻿# Microsoft.CodeAnalysis.Collections
+
+This package contains shared code implementing specialized collection types.
```
Comment (333fred) [inline src/Dependencies/Collections/README.md:3]: If this is going to be shared to other repositories, please doc it explicitly.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/xlf/CompilerExtensionsResources.cs.xlf [hunk 9]
```diff
@@ -17,11 +17,6 @@
         <target state="translated">Element se stejným klíčem, ale odlišnou hodnotou už existuje.</target>
         <note />
       </trans-unit>
-      <trans-unit id="ArgumentException_OtherNotArrayOfCorrectLength">
-        <source>Object is not a array with the same number of elements as the array to compare it to.</source>
-        <target state="translated">Objekt není pole se stejným počtem elementů jako pole, se kterým se má porovnat.</target>
-        <note />
-      </trans-unit>
```
Comment (drewnoakes) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/xlf/CompilerExtensionsResources.cs.xlf:24]: An aside, but in dotnet/project-system we mark `.xlf` files as generated so that they're collapsed by default in GitHub diffs.

https://github.com/dotnet/project-system/commit/3ba729520b8540f4496621b25698b33d36e3cabe

Comment (sharwell) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/xlf/CompilerExtensionsResources.cs.xlf:24]: I mentioned this a long time ago, but it was not adopted for this repository at the time.

File: src/Dependencies/Collections/README.md (hunk 8 above)
Comment (sharwell) [reply src/Dependencies/Collections/README.md:3]: Updated notes, and filed #49991 to address a remaining gap in those notes.

**Review by 333fred — APPROVED** [review]
